File: practice/hackerrank/advanced/counting_tree/CountingTree.java
import java.util.*;

/**
 * Created by bmk on 12/31/16.
 */
public class CountingTree {

    private static Scanner scanner = new Scanner(System.in);
    private static HashSet<Long> duplicatedValues = new HashSet<>();
    private static Node[] nodes;

    public static void main(String[] args) {
        int n = scanner.nextInt();
        int q = scanner.nextInt();
        readTree(n);

        /*for (int i = 0; i < q; i++) {
            int x1 = scanner.nextInt() - 1;
            int y1 = scanner.nextInt() - 1;
            int x2 = scanner.nextInt() - 1;
            int y2 = scanner.nextInt() - 1;

            solve(x1, y1, x2, y2, i+1);
        }*/
    }

    public static void solve(int x1, int y1, int x2, int y2, int run) {
        //TODO: implement me!
    }

    public static void readTree(int n) {
        // read nodes
        nodes = new Node[n];

        for (int i = 0; i < n; i++) {
            long c = scanner.nextLong();
            nodes[i] = new Node(i, c);
        }

        // build tree
        for (int i = 0; i < n - 1; i++) {
            int xId = scanner.nextInt() - 1;
            int yId = scanner.nextInt() - 1;

            Node parent = nodes[xId];
            Node child = nodes[yId];
            child.addParent(parent);
        }

        // add children
        Node root = null;
        for (Node node : nodes) {
            if (node.parent != null) {
                node.parent.addChild(node);
            } else {
                root = node;
            }
        }

        Map<Long, Integer> map = new HashMap<>();
        for (Node node : nodes) {
            if (map.containsKey(node.data)) {
                map.put(node.data, map.get(node.data) + 1);
            } else {
                map.put(node.data, 1);
            }
        }

        for (Long data : map.keySet()) {
            if (map.get(data) > 1) {
                duplicatedValues.add(data);
            }
        }

        List<Node> nextLvlNodes = new ArrayList<>();
        List<HashSet<DupParent>> nextLvlParents = new ArrayList<>();
        nextLvlNodes.add(root);
        nextLvlParents.add(new HashSet<>());

        while (!nextLvlNodes.isEmpty()) {
            List<Node> l = new ArrayList<>();
            List<HashSet<DupParent>> p = new ArrayList<>();

            for (int i = 0; i < nextLvlNodes.size(); i++) {
                Node node = nextLvlNodes.get(i);
                HashSet<DupParent> parents = nextLvlParents.get(i);

                if (duplicatedValues.contains(node.data)) {
                    int idx = 0;
                    for (Node c : node.children) {
                        HashSet<DupParent> copy = new HashSet<>(parents);
                        copy.add(new DupParent(c, idx));
                        idx++;

                        l.add(c);
                        p.add(copy);
                    }
                } else {
                    for (Node c : node.children) {
                        l.add(c);
                        p.add(parents);
                    }
                }
            }

            nextLvlNodes = l;
            nextLvlParents = p;
        }

    }

    public static class Node {
        final int id;
        final long data;
        Node parent;
        HashSet<Node> children;
        HashSet<DupParent> duplicatedParents;

        public Node(int id, long data) {
            this.id = id;
            this.data = data;
            this.parent = null;
            this.children = new HashSet<>();
        }

        public void setDuplicatedParents(HashSet<DupParent> dupParents) {
            this.duplicatedParents = dupParents;
        }

        public void addChild(Node child) {
            this.children.add(child);
        }

        public void addParent(Node parent) {
            if (this.parent != null) {
                List<Node> stack = new ArrayList<>();

                Node p = this;
                while (p != null) {
                    stack.add(p);
                    p = p.parent;
                }

                for (int i = stack.size() - 1; i > 0; i--) {
                    p = stack.get(i);
                    p.parent = stack.get(i - 1);
                }
            }

            this.parent = parent;
        }
    }

    public static class DupParent {
        final Node node;
        final int idx;

        public DupParent(Node node, int idx) {
            this.node = node;
            this.idx = idx;
        }
    }
}

File: practice/hackerrank/advanced/counting_tree/counting_tree.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

struct Node {
  int id;
  long data;
  int parentId;
  int height;
  int run;
};

void connect(Node nodes[], Node& child, Node& parent) {
    if (child.parentId != -1) {
        vector<int> stack;

        int i = child.id;
        while (i != -1) {
            stack.push_back(i);
            i = nodes[i].parentId;
        }

        for (int j = stack.size() - 1; j > 0; j--) {
            nodes[stack[j]].parentId = stack[j-1];
        }
    }

    child.parentId = parent.id;
}

void markPath(Node nodes[], int n1, int n2, int run) {
        int h1 = nodes[n1].height;
        int h2 = nodes[n2].height;

	int ticks = 0;
        if (h1 > h2) {
            while (h1 != h2) {
		nodes[n1].run = run;
                n1 = nodes[n1].parentId;
                h1--;
            }
        } else if (h2 > h1) {
            while (h1 != h2) {
		nodes[n2].run = run;
                n2 = nodes[n2].parentId;
                h2--;
            }
        }

        while (n1 != n2) {
            nodes[n1].run = run;
	    nodes[n2].run = run;
            
            n1 = nodes[n1].parentId;
            n2 = nodes[n2].parentId;
        }

        nodes[n1].run = run;
}

void query(Node nodes[], int x1, int y1, int x2, int y2, int run) {
	markPath(nodes, x1, y1, run);
	markPath(nodes, x2, y2, run);
	
	cout << 0 << endl;
}

int main() {
    	int n, q;
    	cin >> n >> q;

	Node nodes[n];

	// read nodes
	for (int i = 0; i < n; i++) {
		long c;
		cin >> c;
		nodes[i].id = i;
		nodes[i].data = c;
		nodes[i].parentId = -1;
		nodes[i].run = -1;	
	}

	// read edges
	for (int i = 0; i < n - 1; i++) {
            	int xId;
            	int yId;
		cin >> xId;
		cin >> yId;

            	connect(nodes, nodes[xId-1], nodes[yId-1]);
        }

	int rootId;
	vector<int> children[n];
	// add children
        for (int i = 0; i < n; i++) {
		if (nodes[i].parentId != -1) {
			children[nodes[i].parentId].push_back(i);		
		} else {
			rootId = i;		
		}	
	}
	
	// add height info
        int height = 1;
        vector<int> nextLvlNodes = children[rootId];
	nodes[rootId].height = 0;

        while (!nextLvlNodes.empty()) {
            vector<int> l;

            for (int i = 0; i < nextLvlNodes.size(); i++) {
                nodes[nextLvlNodes[i]].height = height;
                
		
		l.insert(l.end(), children[nextLvlNodes[i]].begin(), children[nextLvlNodes[i]].end());
            }

            height++;
            nextLvlNodes = l;
        }


	// read queries
	for (int i = 0; i < q; i++) {
		int x1;
            	int y1;
		int x2;
		int y2;
		cin >> x1;
		cin >> y1;
		cin >> x2;
		cin >> y2;

		query(nodes, x1-1, y1-1, x2-1, y2-1, i);
	}
}



